README.md
# bocfel
Bocfel is an interpreter for the Z-machine. https://cspiegel.github.io/bocfel/


NEWS
2.1.1
• Various small bug fixes.

2.1
• Honor the loop chunk in Blorb sounds.
• Support bleeps under Gargoyle.
• Use separate channels for music and sound effects.

2.0.1
• Don’t patch Bureaucracy if timed input isn’t available.
• Properly set CXXFLAGS if ICU support is requested.
• Avoid a potential buffer overflow in memory I/O.
• Work around a bug in Inform’s “box” statement.

2.0
• Bocfel has been ported from C to C++. This should be transparent for users,
  but is the reason for the major version bump.
• ICU can be used to perform Unicode lowercasing.

1.4.2
• Support auxiliary files (see §7.6 of the Z-Machine Standards Document 1.1).
• Add a patch to properly support delays in Bureaucracy.

1.4.1
• Continue searching for Blorb files if the first one found is invalid.
• Allow patches to be set via the configuration file.
• A new “/config” meta command was added which allows users to edit the
  configuration file with their preferred text editor.

1.4
• A new “/notes” meta command was added which allows users to add arbitrary
  notes to game sessions; these notes are included in save files automatically
  so they are available across game sessions.
• A new persistent transcripting feature was added: if enabled, a transcript of
  every game session is automatically created and included in save files, so it
  is available across game sessions. This can be exported at any time.
• The status line for Planetfall and Stationfall now properly displays the time.
• Non-ASCII characters are again accepted in the Software License Application in
  Bureaucracy.

1.3.2
• Support queued sound effects as described in §9 of The Z-Machine Standards
  Document 1.1.

1.3.1
• Fix two sound bugs: fading in Sherlock works better now, and The Lurking
  Horror no longer crashes when playing some sounds.
• Add support for repeating sounds in The Lurking Horror.

1.3
• Add experimental autosave (disabled by default): Bocfel can now pick up where
  it left off after a previous shutdown.
• Add mouse support.
• Add support for the “routine” operand to @sound_effect.
• Add support for DOS (32-bit only at the moment).
• The XDG Base Directory Specification is now honored on Unix systems.
• Add more patches for Beyond Zork to improve its behavior if the interpreter is
  DOS or Apple II.

1.2.1
• Various small bug fixes.

1.2
• Rework meta saves to fix corner-case issues and improve screen management.
• Add history playback to save files to provide context on restore.
• Update @set_font handling to be in line with the latest standard revision.
• Several small fixes and updates (including but not limited to fixes for issues
  384, 385, 395, and 399 at https://github.com/garglk/garglk/issues/).

1.1
• Support Blorb multimedia resources in non-Gargoyle Glk implementations.
• Fix two potential crashes discovered by American Fuzzy Lop.
• Improved terminal (non-Glk) support: Windows can now do colors and styles, and
  Unix now supports true color on terminals capable of it.
• Fix preloaded input in Beyond Zork.

1.0.1
• Ensure null termination of a string.

1.0
• Add the ability to patch known bugs in games at runtime.
• The username can now be set.

0.9
• UTF-8 is now used for all transcripts and command scripts.
• Watchpoints can now be set through the debugging interface.
• The upper window now tracks resizes properly even when it is closed.

0.6.3.2
• Fix detection of Sherlock version 26-880127.

0.6.3.1
• Fix a potential use of uninitialized variable.

0.6.3
• Improved handling of abbreviations (x, g, z, and o).
• The “attribute 48” bug of Sherlock is now properly handled.
• Blorb sounds are now properly loaded if the story file itself is wrapped in a
  Blorb file.
• Support for “debugging” meta commands, the main use of which is to help find
  cheats without needing to know details of the Z-machine architecture.

0.6.2
• Enable /undo in V5 games which do not support undo (e.g. Hitchhiker’s Guide).
• Add new meta commands /say (allowing text strings which begin with / to be
  sent to the game) and /disable (which disables meta commands for the rest of
  the game session).
• Add new meta commands /ps and /pop which allow in-memory saves to be created,
  providing arbitrary save points.
• Meta commands are no longer included in transcripts.

0.6.1
• Bocfel is now dual-licensed under both the GPLv2 and GPLv3.
• Small bugfixes and general cleanup.

0.6.0
• The configuration options script_on and script_name have been renamed to
  record_on and record_name.
• Support for some meta commands (/undo, /replay, etc.) has been added.
• Assorted minor bugfixes and features.

0.5.5
• Preliminary @sound_effect support.
• Small bugfixes and general cleanup.

0.5.4
• Fix a rather obscure bug triggered when @restart is called inside an
  interrupt.
• Improved resilience in the face of utterly broken story files.
• Minor improvements to the syntax for cheats.
• Improved handling of the configuration file on Windows.

0.5.3
• The initial random seed can now be read from a device such as /dev/random.
• Support for Z-machine versions 1 and 2 is now built unconditionally.
• Small bugfixes and general cleanup.

0.5.2
• Fix a display problem when a newline is printed at the right edge of the upper
  window.
• Provide better diagnostics when a save file cannot be loaded.

0.5.1
• Bocfel now reports itself as a Standard 1.1-compliant interpreter.
• The stack is now properly saved in a @save_undo call.
• Bocfel generates error messages on a few rather questionable uses of the
  Z-machine; some of these instances have been modified to return sensible
  results instead of aborting. The praxix and strictz tests are now passed.

0.5.0
• Initial release.


util.c
/*-
 * Copyright 2010-2012 Chris Spiegel.
 *
 * This file is part of Bocfel.
 *
 * Bocfel is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version
 * 2 or 3, as published by the Free Software Foundation.
 *
 * Bocfel is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Bocfel.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdarg.h>

#include "util.h"
#include "screen.h"
#include "unicode.h"
#include "zterp.h"

#ifdef ZTERP_GLK
#include <glk.h>
#endif

#ifndef ZTERP_NO_SAFETY_CHECKS
unsigned long zassert_pc;

void assert_fail(const char *fmt, ...)
{
  va_list ap;
  char str[1024];

  va_start(ap, fmt);
  vsnprintf(str, sizeof str, fmt, ap);
  va_end(ap);

  snprintf(str + strlen(str), sizeof str - strlen(str), " (pc = 0x%lx)", zassert_pc);

  die("%s", str);
}
#endif

void warning(const char *fmt, ...)
{
  va_list ap;
  char str[1024];

  va_start(ap, fmt);
  vsnprintf(str, sizeof str, fmt, ap);
  va_end(ap);

  show_message("WARNING: %s", str);
}

void die(const char *fmt, ...)
{
  va_list ap;
  char str[1024];

  va_start(ap, fmt);
  vsnprintf(str, sizeof str, fmt, ap);
  va_end(ap);

  show_message("fatal error: %s", str);

#ifdef ZTERP_GLK
#ifdef GARGLK
  fprintf(stderr, "%s\n", str);
#endif
  glk_exit();
#endif

  exit(EXIT_FAILURE);
}

/* This is not POSIX compliant, but it gets the job done.
 * It should not be called more than once.
 */
static int zoptind = 0;
static const char *zoptarg;
static int zgetopt(int argc, char **argv, const char *optstring)
{
  static const char *p = "";
  const char *optp;
  int c;

  if(*p == 0)
  {
    /* No more arguments. */
    if(++zoptind >= argc) return -1;

    p = argv[zoptind];

    /* No more options. */
    if(p[0] != '-' || p[1] == 0) return -1;

    /* Handle “--” */
    if(*++p == '-')
    {
      zoptind++;
      return -1;
    }
  }

  c = *p++;

  optp = strchr(optstring, c);
  if(optp == NULL) return '?';

  if(optp[1] == ':')
  {
    if(*p != 0) zoptarg = p;
    else        zoptarg = argv[++zoptind];

    p = "";
    if(zoptarg == NULL) return '?';
  }

  return c;
}

char *xstrdup(const char *s)
{
  size_t n;
  char *r;

  n = strlen(s) + 1;

  r = malloc(n);
  if(r != NULL) memcpy(r, s, n);

  return r;
}

int process_arguments(int argc, char **argv)
{
  int c;

  while( (c = zgetopt(argc, argv, "a:A:cCdDeE:fFgGiklLmn:N:rR:sS:tT:u:UvxXyYz:Z:")) != -1 )
  {
    switch(c)
    {
      case 'a':
        options.eval_stack_size = strtol(zoptarg, NULL, 10);
        break;
      case 'A':
        options.call_stack_size = strtol(zoptarg, NULL, 10);
        break;
      case 'c':
        options.disable_color = 1;
        break;
      case 'C':
        options.disable_config = 1;
        break;
      case 'd':
        options.disable_timed = 1;
        break;
      case 'D':
        options.disable_sound = 1;
        break;
      case 'e':
        options.enable_escape = 1;
        break;
      case 'E':
        options.escape_string = xstrdup(zoptarg);
        break;
      case 'f':
        options.disable_fixed = 1;
        break;
      case 'F':
        options.assume_fixed = 1;
        break;
      case 'g':
        options.disable_graphics_font = 1;
        break;
      case 'G':
        options.enable_alt_graphics = 1;
        break;
      case 'i':
        options.show_id = 1;
        break;
      case 'k':
        options.disable_term_keys = 1;
        break;
      case 'l':
        options.disable_utf8 = 1;
        break;
      case 'L':
        options.force_utf8 = 1;
        break;
      case 'm':
        options.disable_meta_commands = 1;
        break;
      case 'n':
        options.int_number = strtol(zoptarg, NULL, 10);
        break;
      case 'N':
        options.int_version = zoptarg[0];
        break;
      case 'r':
        options.replay_on = 1;
        break;
      case 'R':
        options.replay_name = xstrdup(zoptarg);
        break;
      case 's':
        options.record_on = 1;
        break;
      case 'S':
        options.record_name = xstrdup(zoptarg);
        break;
      case 't':
        options.transcript_on = 1;
        break;
      case 'T':
        options.transcript_name = xstrdup(zoptarg);
        break;
      case 'u':
        options.max_saves = strtol(zoptarg, NULL, 10);
        break;
      case 'U':
        options.disable_undo_compression = 1;
        break;
      case 'v':
        options.show_version = 1;
        break;
      case 'x':
        options.disable_abbreviations = 1;
        break;
      case 'X':
        options.enable_censorship = 1;
        break;
      case 'y':
        options.overwrite_transcript = 1;
        break;
      case 'Y':
        options.override_undo = 1;
        break;
      case 'z':
        options.random_seed = strtol(zoptarg, NULL, 10);
        break;
      case 'Z':
        options.random_device = xstrdup(zoptarg);
        break;
      default:
        return 0;
    }
  }

  /* Just ignore excess stories for now. */
  if(zoptind < argc) game_file = argv[zoptind];

  return 1;
}


zterp.h
#ifndef ZTERP_H
#define ZTERP_H

#include <stdint.h>

struct options
{
  long eval_stack_size;
  long call_stack_size;
  int disable_color;
  int disable_config;
  int disable_timed;
  int disable_sound;
  int enable_escape;
  char *escape_string;
  int disable_fixed;
  int assume_fixed;
  int disable_graphics_font;
  int enable_alt_graphics;
  int show_id;
  int disable_term_keys;
  int disable_utf8;
  int force_utf8;
  int disable_meta_commands;
  long int_number;
  int int_version;
  int replay_on;
  char *replay_name;
  int record_on;
  char *record_name;
  int transcript_on;
  char *transcript_name;
  long max_saves;
  int disable_undo_compression;
  int show_version;
  int disable_abbreviations;
  int enable_censorship;
  int overwrite_transcript;
  int override_undo;
  long random_seed;
  char *random_device;
};

extern const char *game_file;
extern struct options options;

/* v3 */
#define FLAGS1_STATUSTYPE	(1U << 1)
#define FLAGS1_STORYSPLIT	(1U << 2)
#define FLAGS1_CENSOR		(1U << 3)
#define FLAGS1_NOSTATUS		(1U << 4)
#define FLAGS1_SCREENSPLIT	(1U << 5)
#define FLAGS1_VARIABLE		(1U << 6)

/* v4 */
#define FLAGS1_COLORS		(1U << 0)
#define FLAGS1_PICTURES		(1U << 1)
#define FLAGS1_BOLD		(1U << 2)
#define FLAGS1_ITALIC		(1U << 3)
#define FLAGS1_FIXED		(1U << 4)
#define FLAGS1_SOUND		(1U << 5)
#define FLAGS1_TIMED		(1U << 7)

#define FLAGS2_TRANSCRIPT	(1U << 0)
#define FLAGS2_FIXED		(1U << 1)
#define FLAGS2_STATUS		(1U << 2)
#define FLAGS2_PICTURES		(1U << 3)
#define FLAGS2_UNDO		(1U << 4)
#define FLAGS2_MOUSE		(1U << 5)
#define FLAGS2_COLORS		(1U << 6)
#define FLAGS2_SOUND		(1U << 7)
#define FLAGS2_MENUS		(1U << 8)

#define STATUS_IS_TIME()	(zversion == 3 && (BYTE(0x01) & FLAGS1_STATUSTYPE))
#define TIMER_AVAILABLE()	(zversion >= 4 && (BYTE(0x01) & FLAGS1_TIMED))

struct header
{
  uint16_t release;
  uint16_t dictionary;
  uint16_t objects;
  uint16_t globals;
  uint16_t static_start;
  uint16_t static_end;
  uint16_t abbr;
  uint32_t file_length;
  uint8_t  serial[6];
  uint16_t checksum;
  uint32_t R_O;
  uint32_t S_O;
};

extern uint32_t pc;
extern int zversion;
extern struct header header;
extern uint8_t atable[];

int is_story(const char *);

void write_header(void);

uint32_t unpack(uint16_t, int);
void store(uint16_t);
void process_story(void);

#ifndef ZTERP_NO_CHEAT
int cheat_find_freezew(uint32_t, uint16_t *);
#endif

void znop(void);
void zquit(void);
void zverify(void);
void zpiracy(void);
void zsave5(void);
void zrestore5(void);
void zsound_effect(void);

#endif


memory.c
/*-
 * Copyright 2010-2012 Chris Spiegel.
 *
 * This file is part of Bocfel.
 *
 * Bocfel is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version
 * 2 or 3, as published by the Free Software Foundation.
 *
 * Bocfel is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Bocfel.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdlib.h>
#include <stdint.h>

#include "memory.h"
#include "branch.h"
#include "process.h"
#include "screen.h"
#include "util.h"
#include "zterp.h"

uint8_t *memory, *dynamic_memory;
uint32_t memory_size;

void user_store_byte(uint16_t addr, uint8_t v)
{
  /* If safety checks are off, there’s no point in checking these
   * special cases. */
#ifndef ZTERP_NO_SAFETY_CHECKS
#ifdef ZTERP_TANDY
  if(addr == 0x01)
  {
    ZASSERT(v == BYTE(addr) || (BYTE(addr) ^ v) == 8, "not allowed to modify any bits but 3 at 0x0001");
  }
  else
#endif

  /* 0x10 can’t be modified, but let it slide if the story is storing
   * the same value that’s already there.  This is useful because the
   * flags at 0x10 are stored in a word, so the story possibly could use
   * @storew at 0x10 to modify the bits in 0x11.
   */
  if(addr == 0x10 && BYTE(addr) == v)
  {
    return;
  }
  else
#endif

  if(addr == 0x11)
  {
    ZASSERT((BYTE(addr) ^ v) < 8, "not allowed to modify bits 3-7 at 0x0011");

    if(!output_stream((v & FLAGS2_TRANSCRIPT) ? OSTREAM_SCRIPT : -OSTREAM_SCRIPT, 0)) v &= ~FLAGS2_TRANSCRIPT;

    header_fixed_font = v & FLAGS2_FIXED;
    set_current_style();
  }

  else
  {
    ZASSERT(addr >= 0x40 && addr < header.static_start, "attempt to write to read-only address 0x%lx", (unsigned long)addr);
  }

  STORE_BYTE(addr, v);
}

void user_store_word(uint16_t addr, uint16_t v)
{
  user_store_byte(addr + 0, v >> 8);
  user_store_byte(addr + 1, v & 0xff);
}

void zcopy_table(void)
{
  uint16_t first = zargs[0], second = zargs[1], size = zargs[2];

  if(second == 0)
  {
    for(uint16_t i = 0; i < size; i++) user_store_byte(first + i, 0);
  }
  else if( (first > second) || (int16_t)size < 0 )
  {
    long n = labs((int16_t)size);
    for(long i = 0; i < n; i++) user_store_byte(second + i, user_byte(first + i));
  }
  else
  {
    for(uint16_t i = 0; i < size; i++) user_store_byte(second + size - i - 1, user_byte(first + size - i - 1));
  }
}

void zscan_table(void)
{
  uint16_t addr = zargs[1];

  if(znargs < 4) zargs[3] = 0x82;

  for(uint16_t i = 0; i < zargs[2]; i++)
  {
    if(
        ( (zargs[3] & 0x80) && (user_word(addr) == zargs[0])) ||
        (!(zargs[3] & 0x80) && (user_byte(addr) == zargs[0]))
      )
    {
      store(addr);
      branch_if(1);
      return;
    }

    addr += zargs[3] & 0x7f;
  }

  store(0);
  branch_if(0);
}

void zloadw(void)
{
  store(user_word(zargs[0] + (2 * zargs[1])));
}

void zloadb(void)
{
  store(user_byte(zargs[0] + zargs[1]));
}

void zstoreb(void)
{
  user_store_byte(zargs[0] + zargs[1], zargs[2]);
}

void zstorew(void)
{
  user_store_word(zargs[0] + (2 * zargs[1]), zargs[2]);
}


process.c
/*-
 * Copyright 2010-2012 Chris Spiegel.
 *
 * This file is part of Bocfel.
 *
 * Bocfel is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version
 * 2 or 3, as published by the Free Software Foundation.
 *
 * Bocfel is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Bocfel.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdlib.h>
#include <stdint.h>
#include <stddef.h>
#include <setjmp.h>

#ifdef ZTERP_GLK
#include <glk.h>
#endif

#include "process.h"
#include "branch.h"
#include "dict.h"
#include "math.h"
#include "memory.h"
#include "objects.h"
#include "random.h"
#include "screen.h"
#include "stack.h"
#include "util.h"
#include "zoom.h"
#include "zterp.h"

uint16_t zargs[8];
int znargs;

static jmp_buf *jumps;
static size_t njumps;

/* Each time an interrupt happens, process_instructions() is called
 * (effectively starting a whole new round of interpreting).  This
 * variable holds the current level of interpreting: 0 for no
 * interrupts, 1 if one interrupt has been called, 2 if an interrupt was
 * called inside of an interrupt, and so on.
 */
static long ilevel = -1;

long interrupt_level(void)
{
  return ilevel;
}

/* When this is called, the interrupt at level “level” will stop
 * running: if a single interrupt is running, then break_from(1) will
 * stop the interrupt, going back to the main program.  Breaking from
 * interrupt level 0 (which is not actually an interrupt) will end the
 * program.  This is how @quit is implemented.
 */
void break_from(long level)
{
  ilevel = level - 1;
  longjmp(jumps[level], 1);
}

/* To signal a restart, longjmp() is called with 2; this advises
 * process_instructions() to restart the story file and then continue
 * execution, whereas a value of 1 tells it to return immediately.
 */
static void zrestart(void)
{
  ilevel = 0;
  longjmp(jumps[0], 2);
}

/* If a restore happens inside of an interrupt, the level needs to be
 * set back to 0; to differentiate from a restart, call longjmp() with a
 * value of 3.
 */
void reset_level(void)
{
  ilevel = 0;
  longjmp(jumps[0], 3);
}

/* Returns 1 if decoded, 0 otherwise (omitted) */
static int decode_base(uint8_t type, uint16_t *loc)
{
  switch(type)
  {
    case 0: /* Large constant. */
      *loc = WORD(pc);
      pc += 2;
      break;
    case 1: /* Small constant. */
      *loc = BYTE(pc++);
      break;
    case 2: /* Variable. */
      *loc = variable(BYTE(pc++));
      break;
    default: /* Omitted. */
      return 0;
  }

  return 1;
}

static void decode_var(uint8_t types)
{
  uint16_t ret;

  for(int i = 6; i >= 0; i -= 2)
  {
    if(!decode_base((types >> i) & 0x03, &ret)) return;
    zargs[znargs++] = ret;
  }
}

/* op[0] is 0OP, op[1] is 1OP, etc */
static void (*op[5][256])(void);
static const char *opnames[5][256];
enum { ZERO, ONE, TWO, VAR, EXT };

#define op_call(opt, opnumber)	(op[opt][opnumber]())

/* This nifty trick is from Frotz. */
static void zextended(void)
{
  uint8_t opnumber = BYTE(pc++);

  decode_var(BYTE(pc++));

  /* §14.2.1
   * The exception for 0x80–0x83 is for the Zoom extensions.
   * Standard 1.1 implicitly updates §14.2.1 to recommend ignoring
   * opcodes in the range EXT:30 to EXT:255, due to the fact that
   * @buffer_screen is EXT:29.
   */
  if(opnumber > 0x1d && (opnumber < 0x80 || opnumber > 0x83)) return;

  op_call(EXT, opnumber);
}

static void illegal_opcode(void)
{
#ifndef ZTERP_NO_SAFETY_CHECKS
  die("illegal opcode (pc = 0x%lx)", zassert_pc);
#else
  die("illegal opcode");
#endif
}

void setup_opcodes(void)
{
  for(int i = 0; i < 5; i++)
  {
    for(int j = 0; j < 256; j++)
    {
      op[i][j] = illegal_opcode;
    }
  }
#define OP(minver, maxver, args, opcode, fn) do { if(zversion >= minver && zversion <= maxver) op[args][opcode] = fn; opnames[args][opcode] = #fn; } while(0)
  OP(1, 6, ZERO, 0x00, zrtrue);
  OP(1, 6, ZERO, 0x01, zrfalse);
  OP(1, 6, ZERO, 0x02, zprint);
  OP(1, 6, ZERO, 0x03, zprint_ret);
  OP(1, 6, ZERO, 0x04, znop);
  OP(1, 4, ZERO, 0x05, zsave);
  OP(1, 4, ZERO, 0x06, zrestore);
  OP(1, 6, ZERO, 0x07, zrestart);
  OP(1, 6, ZERO, 0x08, zret_popped);
  OP(1, 4, ZERO, 0x09, zpop);
  OP(5, 6, ZERO, 0x09, zcatch);
  OP(1, 6, ZERO, 0x0a, zquit);
  OP(1, 6, ZERO, 0x0b, znew_line);
  OP(3, 3, ZERO, 0x0c, zshow_status);
  OP(4, 6, ZERO, 0x0c, znop); /* §15: Technically illegal in V4+, but a V5 Wishbringer accidentally uses this opcode. */
  OP(3, 6, ZERO, 0x0d, zverify);
  OP(5, 6, ZERO, 0x0e, zextended);
  OP(5, 6, ZERO, 0x0f, zpiracy);

  OP(1, 6, ONE, 0x00, zjz);
  OP(1, 6, ONE, 0x01, zget_sibling);
  OP(1, 6, ONE, 0x02, zget_child);
  OP(1, 6, ONE, 0x03, zget_parent);
  OP(1, 6, ONE, 0x04, zget_prop_len);
  OP(1, 6, ONE, 0x05, zinc);
  OP(1, 6, ONE, 0x06, zdec);
  OP(1, 6, ONE, 0x07, zprint_addr);
  OP(4, 6, ONE, 0x08, zcall_1s);
  OP(1, 6, ONE, 0x09, zremove_obj);
  OP(1, 6, ONE, 0x0a, zprint_obj);
  OP(1, 6, ONE, 0x0b, zret);
  OP(1, 6, ONE, 0x0c, zjump);
  OP(1, 6, ONE, 0x0d, zprint_paddr);
  OP(1, 6, ONE, 0x0e, zload);
  OP(1, 4, ONE, 0x0f, znot);
  OP(5, 6, ONE, 0x0f, zcall_1n);

  OP(1, 6, TWO, 0x01, zje);
  OP(1, 6, TWO, 0x02, zjl);
  OP(1, 6, TWO, 0x03, zjg);
  OP(1, 6, TWO, 0x04, zdec_chk);
  OP(1, 6, TWO, 0x05, zinc_chk);
  OP(1, 6, TWO, 0x06, zjin);
  OP(1, 6, TWO, 0x07, ztest);
  OP(1, 6, TWO, 0x08, zor);
  OP(1, 6, TWO, 0x09, zand);
  OP(1, 6, TWO, 0x0a, ztest_attr);
  OP(1, 6, TWO, 0x0b, zset_attr);
  OP(1, 6, TWO, 0x0c, zclear_attr);
  OP(1, 6, TWO, 0x0d, zstore);
  OP(1, 6, TWO, 0x0e, zinsert_obj);
  OP(1, 6, TWO, 0x0f, zloadw);
  OP(1, 6, TWO, 0x10, zloadb);
  OP(1, 6, TWO, 0x11, zget_prop);
  OP(1, 6, TWO, 0x12, zget_prop_addr);
  OP(1, 6, TWO, 0x13, zget_next_prop);
  OP(1, 6, TWO, 0x14, zadd);
  OP(1, 6, TWO, 0x15, zsub);
  OP(1, 6, TWO, 0x16, zmul);
  OP(1, 6, TWO, 0x17, zdiv);
  OP(1, 6, TWO, 0x18, zmod);
  OP(4, 6, TWO, 0x19, zcall_2s);
  OP(5, 6, TWO, 0x1a, zcall_2n);
  OP(5, 6, TWO, 0x1b, zset_colour);
  OP(5, 6, TWO, 0x1c, zthrow);

  OP(1, 6, VAR, 0x00, zcall);
  OP(1, 6, VAR, 0x01, zstorew);
  OP(1, 6, VAR, 0x02, zstoreb);
  OP(1, 6, VAR, 0x03, zput_prop);
  OP(1, 6, VAR, 0x04, zread);
  OP(1, 6, VAR, 0x05, zprint_char);
  OP(1, 6, VAR, 0x06, zprint_num);
  OP(1, 6, VAR, 0x07, zrandom);
  OP(1, 6, VAR, 0x08, zpush);
  OP(1, 6, VAR, 0x09, zpull);
  OP(3, 6, VAR, 0x0a, zsplit_window);
  OP(3, 6, VAR, 0x0b, zset_window);
  OP(4, 6, VAR, 0x0c, zcall_vs2);
  OP(4, 6, VAR, 0x0d, zerase_window);
  OP(4, 6, VAR, 0x0e, zerase_line);
  OP(4, 6, VAR, 0x0f, zset_cursor);
  OP(4, 6, VAR, 0x10, zget_cursor);
  OP(4, 6, VAR, 0x11, zset_text_style);
  OP(4, 6, VAR, 0x12, znop); /* XXX buffer_mode */
  OP(3, 6, VAR, 0x13, zoutput_stream);
  OP(3, 6, VAR, 0x14, zinput_stream);
  OP(3, 6, VAR, 0x15, zsound_effect);
  OP(4, 6, VAR, 0x16, zread_char);
  OP(4, 6, VAR, 0x17, zscan_table);
  OP(5, 6, VAR, 0x18, znot);
  OP(5, 6, VAR, 0x19, zcall_vn);
  OP(5, 6, VAR, 0x1a, zcall_vn2);
  OP(5, 6, VAR, 0x1b, ztokenise);
  OP(5, 6, VAR, 0x1c, zencode_text);
  OP(5, 6, VAR, 0x1d, zcopy_table);
  OP(5, 6, VAR, 0x1e, zprint_table);
  OP(5, 6, VAR, 0x1f, zcheck_arg_count);

  OP(5, 6, EXT, 0x00, zsave5);
  OP(5, 6, EXT, 0x01, zrestore5);
  OP(5, 6, EXT, 0x02, zlog_shift);
  OP(5, 6, EXT, 0x03, zart_shift);
  OP(5, 6, EXT, 0x04, zset_font);
  OP(6, 6, EXT, 0x05, znop); /* XXX draw_picture */
  OP(6, 6, EXT, 0x06, zpicture_data);
  OP(6, 6, EXT, 0x07, znop); /* XXX erase_picture */
  OP(6, 6, EXT, 0x08, znop); /* XXX set_margins */
  OP(5, 6, EXT, 0x09, zsave_undo);
  OP(5, 6, EXT, 0x0a, zrestore_undo);
  OP(5, 6, EXT, 0x0b, zprint_unicode);
  OP(5, 6, EXT, 0x0c, zcheck_unicode);
  OP(5, 6, EXT, 0x0d, zset_true_colour);
  OP(6, 6, EXT, 0x10, znop); /* XXX move_window */
  OP(6, 6, EXT, 0x11, znop); /* XXX window_size */
  OP(6, 6, EXT, 0x12, znop); /* XXX window_style */
  OP(6, 6, EXT, 0x13, zget_wind_prop);
  OP(6, 6, EXT, 0x14, znop); /* XXX scroll_window */
  OP(6, 6, EXT, 0x15, zpop_stack);
  OP(6, 6, EXT, 0x16, znop); /* XXX read_mouse */
  OP(6, 6, EXT, 0x17, znop); /* XXX mouse_window */
  OP(6, 6, EXT, 0x18, zpush_stack);
  OP(6, 6, EXT, 0x19, znop); /* XXX put_wind_prop */
  OP(6, 6, EXT, 0x1a, zprint_form);
  OP(6, 6, EXT, 0x1b, zmake_menu);
  OP(6, 6, EXT, 0x1c, znop); /* XXX picture_table */
  OP(6, 6, EXT, 0x1d, zbuffer_screen);

  /* Zoom extensions. */
  OP(5, 6, EXT, 0x80, zstart_timer);
  OP(5, 6, EXT, 0x81, zstop_timer);
  OP(5, 6, EXT, 0x82, zread_timer);
  OP(5, 6, EXT, 0x83, zprint_timer);
#undef OP
}

void process_instructions(void)
{
  if(njumps <= ++ilevel)
  {
    jumps = realloc(jumps, ++njumps * sizeof *jumps);
    if(jumps == NULL) die("unable to allocate memory for jump buffer");
  }

  switch(setjmp(jumps[ilevel]))
  {
    case 1: /* Normal break from interrupt. */
      return;
    case 2: /* Special break: a restart was requested. */
      {
        /* §6.1.3: Flags2 is preserved on a restart. */
        uint16_t flags2 = WORD(0x10);

        process_story();

        STORE_WORD(0x10, flags2);
      }
      break;
    case 3: /* Do nothing: restore was called so keep interpreting. */
      break;
  }

  while(1)
  {
    uint8_t opcode;

#if defined(ZTERP_GLK) && defined(ZTERP_GLK_TICK)
    glk_tick();
#endif

    ZPC(pc);

    opcode = BYTE(pc++);

    /* long 2OP */
    if(opcode < 0x80)
    {
      znargs = 2;

      if(opcode & 0x40) zargs[0] = variable(BYTE(pc++));
      else              zargs[0] = BYTE(pc++);

      if(opcode & 0x20) zargs[1] = variable(BYTE(pc++));
      else              zargs[1] = BYTE(pc++);

      op_call(TWO, opcode & 0x1f);
    }

    /* short 1OP */
    else if(opcode < 0xb0)
    {
      znargs = 1;

      if(opcode < 0x90) /* large constant */
      {
        zargs[0] = WORD(pc);
        pc += 2;
      }
      else if(opcode < 0xa0) /* small constant */
      {
        zargs[0] = BYTE(pc++);
      }
      else /* variable */
      {
        zargs[0] = variable(BYTE(pc++));
      }

      op_call(ONE, opcode & 0x0f);
    }

    /* short 0OP (plus EXT) */
    else if(opcode < 0xc0)
    {
      znargs = 0;

      op_call(ZERO, opcode & 0x0f);
    }

    /* variable 2OP */
    else if(opcode < 0xe0)
    {
      znargs = 0;

      decode_var(BYTE(pc++));

      op_call(TWO, opcode & 0x1f);
    }

    /* Double variable VAR */
    else if(opcode == 0xec || opcode == 0xfa)
    {
      uint8_t types1, types2;

      znargs = 0;

      types1 = BYTE(pc++);
      types2 = BYTE(pc++);
      decode_var(types1);
      decode_var(types2);

      op_call(VAR, opcode & 0x1f);
    }

    /* variable VAR */
    else
    {
      znargs = 0;

      read_pc = pc - 1;

      decode_var(BYTE(pc++));

      op_call(VAR, opcode & 0x1f);
    }
  }
}
